### PR TITLE
Create a wrapper for metasm's C compiling function

### PR DESCRIPTION
**Pull request by wchen-r7**
This is a wrapper for metasm's C compiling function. The purpose is to be able to write something that doesn't require you to define the APIs every single time from scratch, which is time consuming and an unpleasant experience.

windows.h comes first, the rest will come later (as separate PRs) if this is successfully landed.

## Verification

- [x] Make sure Travis is green.
- [x] Save the following c code as `template.c`:

```
#include <Windows.h>

int main(void) {
  MessageBox(NULL, "Hello world", "Hello", MB_OK);
  return 0;
}
```

- [x] Start msfconsole, and switch to `irb`
- [x] Enter: `require 'metasploit/framework/compiler/windows'`
- [x] Enter: `exe = Metasploit::Framework::Compiler::Windows.compile_c(File.read('template.c'))`
- [x] Enter: `File.write('test.exe', exe)`
- [x] Move the test.exe file to a Windows machine, execute it. You should get a "Hello world" prompt.

### CODE REVIEW

File: data/headers/win32/Windows.h [hunk 1]
```diff
@@ -0,0 +1,533 @@
+#define MAX_PATH 260
```
Comment (busterb) [inline data/headers/win32/Windows.h]: We should include the Metasploit license header on top of this, so it's clear we're not actually redistributing Microsoft's headers, but instead created new ones from scratch.

File: data/headers/win32/stddef.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
+#define NULL ((void *)0)
```
Comment (busterb) [inline data/headers/win32/stddef.h]: Same license note here.

File: lib/metasploit/framework/compiler/win32.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+      class Win32
+
+        # Returns the binary of a compiled source.
+        #
+        # @param c_template [String] The C source code to compile.
+        # @param type [Symbol] PE type, either :exe or :dll
+        # @raise [NotImplementedError] If the type is not supported.
+        # @return [String] The compiled code.
+        def self.compile(c_template, type=:exe)
```
Comment (busterb) [inline lib/metasploit/framework/compiler/win32.rb]: this looks good. I like that the Metasm bits are abstracted, so we could substitute other compilers as well.

However, I'd note that there are other Windows compilers we may want to support in the future (C#, etc.). Should we call this function 'compile_c' with a mind to the future?

File: lib/metasploit/framework/compiler/win32.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            raise NotImplementedError
+          end
+        end
+
+        # Saves the compiled code as a file. This is basically a wrapper of #self.compile.
+        #
+        # @param out_file [String] The file path to save the binary as.
+        # @param c_template [String] The C source code to compile.
+        # @param type [Symbol] PE type, either :exe or :dll
+        # @return [Integer] The number of bytes written.
```
Comment (busterb) [inline lib/metasploit/framework/compiler/win32.rb]: Maybe rename this to 'compile_asm' as well, since 'as' doesn't scream 'assembly' to me.

**Comment by busterb** [comment]
I only had minor comments, this looks great!

**Comment by busterb** [comment]
Can I generate 64-bit shellcode with Metasm as well? I'm curious if 'Win32' is clear enough for the module name or if we should call it 'Win32_x86' so we can add 'Win32_x64' later (or does an architecture flag work better in the method signature?)

**Comment by wchen-r7** [comment]
It looks like you can definitely build a x86_64 binary with meatsm. I'll update the namespace. Thanks.

**Comment by wchen-r7** [comment]
So instead of `win32`, I changed the namespace to `windows`, and `compile_c` allows the user to specify the `cpu`. If that's not okay, let me know, I'll tweak again. Thanks.

**Comment by wchen-r7** [comment]
My PR is ready.

**Comment by wchen-r7** [comment]
@bcook-r7 Anything I can do to keep the momentum going? Think anybody else should take a look at the PR, too?

**Comment by busterb** [comment]
Let's go for it.

**Comment by busterb** [comment]
## Release Notes

This adds a new Metasploit interface for building shellcode in C using metasm's builtin compiler support combined with builtin headers for interfacing with common Windows Win32 APIs.

**Comment by busterb** [comment]
I think this is definitely a if-you-build-it-they-will come. Looking forward to what's next.

**Pull request merged**